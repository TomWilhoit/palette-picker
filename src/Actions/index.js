export const addProjects = (projects) => ({
  type: 'ADD_PROJECTS',
  projects
})

export const addPalettes = (palettes) => ({
  type: 'ADD_PALETTES',
  palettes
})

export const addCurrentProject = (project) => ({
  type: 'ADD_CURRENT_PROJECT',
  project
})