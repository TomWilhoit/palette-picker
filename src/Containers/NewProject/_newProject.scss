.add-project {
  height: 100%;
  form {
    display: flex;
    height: 100%;
  }
}

.new-project-input {
  @include input;
  margin-left: 10px;
  width: 78%;
}

.add-project-btn {
  @include small-btn;
  background-color: $add;
  padding: 0;
  width: 20%;
  .fa-plus {
    color: #d4d8d0;
  }
}

.project-error {
  display: block;
}