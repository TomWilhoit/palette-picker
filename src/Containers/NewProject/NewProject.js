import React from "react";
import { connect } from "react-redux";
import { addProject, updateCurrentProject } from "../../Actions/index";
import { apiCall, createOptions } from "../../Utils/API";
import PropTypes from "prop-types";

export class NewProject extends React.Component {
  constructor(props) {
    super(props);
    this.state = { 
      name: "",
   };
  }

  handleChange = e => {
    this.setState({ name: e.target.value });
  }

  handleClick = e => {
    e.preventDefault();
    this.addNewProject();
  }

  addNewProject = async () => {
    const { name } = this.state;
    if (!name) {
      this.props.setError("Projects must be given a name!");
      return;
    }
    const nameToSend = this.props.checkForSameName(name, "projects");
    const options = createOptions("POST", { name: nameToSend });
    try {
      const response = await apiCall("projects", options);
      this.props.addProject({ name: nameToSend, id: response.id });
      this.selectAddedProject(response.id);
    } catch (error) {
      this.props.setError(`Error: ${error.message}!`);
    }
    this.clearInput();
    this.props.clearError();
  }

  selectAddedProject = id => {
    this.props.updateCurrentProject(id);
  }

  clearInput = () => {
    document.getElementById("newProjectInput").value = "";
    this.setState({ name: "" });
  }

  render() {
    return (
      <div className="add-project">
        <form className="form" onSubmit={this.handleClick}>
          <input
            className="new-project-input"
            placeholder="Project Name..."
            defaultValue={this.state.name}
            onKeyUp={this.handleChange}
            id="newProjectInput"
          />
          <button className="add-project-btn">
            <i className="fas fa-plus" />
          </button>
        </form>
      </div>
    );
  }
}

NewProject.propTypes = {
  projects: PropTypes.array,
  palettes: PropTypes.array,
  checkForSameName: PropTypes.func,
  clearError: PropTypes.func,
  setError: PropTypes.func,
};

export const mapStateToProps = state => ({
  projects: state.projects,
  palettes: state.palettes,
});

export const mapDispatchToProps = dispatch => ({
  addProject: project => dispatch(addProject(project)),
  updateCurrentProject: project => dispatch(updateCurrentProject(project)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(NewProject);