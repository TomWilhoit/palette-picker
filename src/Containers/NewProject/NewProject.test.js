import React from "react";
import { NewProject } from "./NewProject";
import { mapStateToProps } from "./NewProject";
import { mapDispatchToProps } from "./NewProject";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import { addProject } from "../../Actions";

describe("NewProject", () => {
  let wrapper;
  let mockProjects = [{ name: "Tom" }];
  
  beforeEach(() => {
    wrapper = shallow(<NewProject projects={mockProjects} />);
  });

  describe("on load", () => {
    it("should match the snapshot with all data passed in", () => {
      expect(wrapper).toMatchSnapshot();
    })
  
    it("should have default state", () => {
      expect(wrapper.state()).toEqual({
        name: "",
        error: ""
      });
    })
  })

  it("should call handleChange when keyed up", () => {
    const mockEvent = {
      target: "JIM"
    };
    jest.spyOn(wrapper.instance(), "handleChange");
    wrapper.instance().forceUpdate();
    wrapper.find(".new-project-input").simulate("keyUp", mockEvent);
    expect(wrapper.instance().handleChange).toHaveBeenCalled();
  })

  it("should call handleClick when submitted", () => {
    const mockEvent = {
      target: "JIM",
      preventDefault: jest.fn()
    };
    jest.spyOn(wrapper.instance(), "handleClick");
    wrapper.instance().forceUpdate();
    wrapper.find(".form").simulate("submit", mockEvent);
    expect(wrapper.instance().handleClick).toHaveBeenCalled();
  })

  it("should check for repeated names", () => {
    wrapper.setState({ name: "Tom" });
    wrapper.instance().checkForRepeatName();
    expect(wrapper.state().name).toEqual("Tom1");
  })

  it("should keep the same name if name is not repeated", () => {
    wrapper.setState({ name: "UniqueName" });
    wrapper.instance().checkForRepeatName();
    expect(wrapper.state().name).toEqual("UniqueName");
  })

  it.skip("should add a new project", () => {
    let fetchOptions = jest.fn();
    wrapper.setState({ name: "hello" });
    wrapper.instance().addNewProject();
    expect(fetchOptions).toHaveBeenCalled();
    // expect(wrapper.state.error).toEqual("Projects must have a name!")
  })

  it.skip("should show an error message if a user adds a project with no name", () => {
    wrapper.setState({ name: "", error: "" })
    wrapper.instance().addNewProject()
    expect(wrapper.state.error).toEqual("Projects must have a name!")
  })

  it("should mapStateToProps", () => {
    const mockState = {
      projects: [{ name: "Tom" }],
      palettes: [{ name: "Mason", projectId: 4 }],
      currentProject: 4
    };
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(mockState);
  })

  it("should map dispatch to props", () => {
    const mockProject = { name: "Tommy" };
    const mockDispatch = jest.fn();
    const actionToDispatch = addProject(mockProject);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addProject(mockProject);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  })
})

describe("handleChange", () => {

})

describe("handleClick", () => {
  
})

describe("addNewProject", () => {
  
})

describe("selectAddedProject", () => {
  
})

describe("clearInput", () => {
  
})

describe("mapStateToProps", () => {
  
})

describe("mapDispatchToProps", () => {
  
})