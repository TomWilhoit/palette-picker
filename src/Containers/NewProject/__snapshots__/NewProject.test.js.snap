// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewProject on load should match the snapshot with all data passed in 1`] = `
<div
  className="add-project"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      className="new-project-input"
      defaultValue=""
      id="newProjectInput"
      onKeyUp={[Function]}
      placeholder="Project Name..."
    />
    <button
      className="add-project-btn"
    >
      <i
        className="fas fa-plus"
      />
    </button>
  </form>
</div>
`;

exports[`NewProject should match the snapshot with all data passed in 1`] = `
<div
  className="add-project"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      className="new-project-input"
      defaultValue=""
      id="newProjectInput"
      onKeyUp={[Function]}
      placeholder="Project Name..."
    />
    <button
      className="add-project-btn"
    >
      <i
        className="fas fa-plus"
      />
    </button>
  </form>
</div>
`;
